### PR TITLE
UI/client count 1.10

### PR DESCRIPTION
**Pull request by Monkeychip**
This is hidden work for the Client Count. 1.10 feature. See video on the code you uncomment to show the work. Every change has been peer reviewed by Claire or myself. However, it would be nice to get a review outside us two for this first chunk of the feature. 

We want to merge this into main to avoid falling behind on changes. We are also certain (as you can be) that this work will be complete and done for 1.10. 

All the API endpoints are mocked at this point using mirage.

Confirmed no changelog at this point until the final PR.

https://user-images.githubusercontent.com/6618863/148270259-cb91c9f3-ff54-438e-a16e-9a9f5a49ff6e.mp4

### CODE REVIEW

File: ui/lib/core/addon/components/bar-chart.js [hunk 1]
```diff
@@ -47,11 +47,10 @@ const CHAR_LIMIT = 18; // character count limit for y-axis labels to trigger tru
...
 
 // COLOR THEME:
-const BAR_COLOR_DEFAULT = ['#BFD4FF', '#8AB1FF'];
+const LIGHT_AND_DARK_BLUE = ['#BFD4FF', '#8AB1FF'];
 const BAR_COLOR_HOVER = ['#1563FF', '#0F4FD1'];
-const BACKGROUND_BAR_COLOR = '#EBEEF2';
 const TOOLTIP_BACKGROUND = '#525761';
-
+const GREY = '#EBEEF2';
 class BarChartComponent extends Component {
```
Comment (hellobontempo) [inline ui/lib/core/addon/components/bar-chart.js:54]: This component will ultimately be deleted

File: ui/package.json [hunk 2]
```diff
@@ -124,6 +125,7 @@
     "ember-load-initializers": "^2.1.2",
     "ember-maybe-import-regenerator": "^0.1.6",
     "ember-maybe-in-element": "^2.0.3",
+    "ember-modal-dialog": "^3.0.3",
```
Comment (hellobontempo) [inline ui/package.json:128]: This is for the tooltip the charts use

File: ui/package.json [hunk 3]
```diff
@@ -138,6 +139,7 @@
     "ember-svg-jar": "^2.1.0",
     "ember-template-lint": "^3.14.0",
     "ember-test-selectors": "^2.1.0",
+    "ember-tether": "^2.0.1",
```
Comment (hellobontempo) [inline ui/package.json:142]: Required for `ember-modal-dialog`

File: ui/app/utils/chart-helpers.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { format } from 'd3-format';
```
Comment (hashishaw) [inline ui/app/utils/chart-helpers.js:1]: 👏

File: ui/mirage/config.js [hunk 5]
```diff
@@ -22,11 +22,280 @@ export default function () {
     };
   });
 
-  this.get('/sys/internal/ui/feature-flags', (db) => {
-    const featuresResponse = db.features.first();
+  this.get('/sys/internal/ui/feature-flags', () => {
```
Comment (hashishaw) [inline ui/mirage/config.js:25]: I know this is just a mirage file, but this doesn't look like the right path for this response?

File: ui/app/templates/vault/cluster/clients/index.hbs [hunk 6]
```diff
@@ -62,5 +63,6 @@
   </Toolbar>
   <Clients::Config @model={{this.model.config}} @isLoading={{this.currentlyLoading}} />
 {{else}}
+  {{!-- <Clients::Dashboard @tab={{this.tab}} @model={{this.model}} @isLoading={{this.currentlyLoading}} /> --}}
```
Comment (hashishaw) [inline ui/app/templates/vault/cluster/clients/index.hbs:66]: Is this still needed?

File: ui/app/templates/vault/cluster/clients/index.hbs [hunk 7]
```diff
@@ -62,5 +63,6 @@
   </Toolbar>
   <Clients::Config @model={{this.model.config}} @isLoading={{this.currentlyLoading}} />
 {{else}}
```
Comment (hashishaw) [inline ui/app/templates/vault/cluster/clients/index.hbs:65]: It might be cleaner to check for the tab here and render a different component rather than keeping it all in the Clients::Dashboard component. I know that the layouts are very similar, but that component is getting hard to follow and I'd prefer a little duplication over not being able to parse what the template should look like in a given state.

File: ui/app/templates/vault/cluster/clients/index.hbs [hunk 8]
```diff
@@ -7,6 +7,7 @@
 </PageHeader>
 
 <div class="tabs-container box is-bottomless is-marginless is-fullwidth is-paddingless" data-test-pricing-metrics>
+  {{! template-lint-configure no-unknown-arguments-for-builtin-components "warn" }}
```
Comment (hashishaw) [inline ui/app/templates/vault/cluster/clients/index.hbs:10]: Why are we changing this to warn?

File: ui/app/templates/components/clients/horizontal-bar-chart.hbs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<svg
+  class="chart is-horizontal"
+  {{on "mouseleave" this.removeTooltip}}
+  {{did-insert this.renderChart @dataset}}
+  {{did-update this.renderChart @dataset}}
```
Comment (hashishaw) [inline ui/app/templates/components/clients/horizontal-bar-chart.hbs:5]: Are the rest of the charts also going to get `did-update` so they're responsive?

File: ui/app/styles/utils/_spacing.scss [hunk 10]
```diff
@@ -4,6 +4,7 @@ $spacing-xxs: 4px;
 $spacing-xs: 8px;
 $spacing-s: 12px;
 $spacing-m: 16px;
+$spacing-xm: 18px;
```
Comment (hashishaw) [inline ui/app/styles/utils/_spacing.scss:7]: Since this value is only used in charts.scss I'd rather define it there and use it within that file's scope only, instead of mucking up our already-in-flux util spacing. Or maybe define here but as `$spacing-chart`

File: ui/app/components/clients/line-chart.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+ * LineChart components are used to display data in a line plot with accompanying tooltip
+ *
+ * @example
+ * ```js
+ * <LineChart @dataset={dataset} />
+ * ```
+ * @param {array} dataset - dataset is an array of objects
+ */
+
+export default class LineChart extends Component {
```
Comment (hashishaw) [inline ui/app/components/clients/line-chart.js:23]: If the charts share some core things, I think you could make a base `Chart` component and then import & extend that so that if shared interactions/styles update, you only have to update it on the base component

**Review by hashishaw — COMMENTED** [review]
This is a great start! Some non-blocking comments, but it might be best to address them before we merge and forget they exist :)

File: ui/app/utils/chart-helpers.js (hunk 4 above)
Comment (hashishaw) [reply ui/app/utils/chart-helpers.js:1]: (clapping for this util file not the line 😂 )

File: ui/app/templates/vault/cluster/clients/index.hbs (hunk 6 above)
Comment (Monkeychip) [reply ui/app/templates/vault/cluster/clients/index.hbs:66]: Yeah, this is the hidden code. It's how users can access the 1.9 client counts while we work on 1.10 which has no data coming in.

File: ui/app/templates/vault/cluster/clients/index.hbs (hunk 8 above)
Comment (Monkeychip) [reply ui/app/templates/vault/cluster/clients/index.hbs:10]: Warn as kind of a reminder. We likely can't do anything about it soon, but it's not something we want to forget. A discussion Jordan and I had.

File: ui/app/templates/components/clients/horizontal-bar-chart.hbs (hunk 9 above)
Comment (Monkeychip) [reply ui/app/templates/components/clients/horizontal-bar-chart.hbs:5]: Maybe. It depends on the data flow. Right now the data is going to come into the dashboard component and be sent to components from there. But this is a good reminder. I'll check-in with Claire on this.

File: ui/app/templates/vault/cluster/clients/index.hbs (hunk 7 above)
Comment (Monkeychip) [reply ui/app/templates/vault/cluster/clients/index.hbs:65]: Good idea. I'll put a comment in the code as right now changing this will break the layout for client count view 1.9.

File: ui/app/templates/components/clients/horizontal-bar-chart.hbs (hunk 9 above)
Comment (hellobontempo) [reply ui/app/templates/components/clients/horizontal-bar-chart.hbs:5]: Yes - thank you for pointing out! It will be added to the other charts once we have the calendar wired up, which will update the dataset passed in. (Since this code was taken from the bar chart in 1.9 it already has `did-update`. I didn't want to blindly include the helper in the other charts until we had a way to mock the dataset changing and test the functionality)

File: ui/app/components/clients/line-chart.js (hunk 11 above)
Comment (hellobontempo) [reply ui/app/components/clients/line-chart.js:23]: Can we sync on this to clarify? I think I follow, but want to make sure I'm understanding!

File: ui/mirage/config.js (hunk 5 above)
Comment (hellobontempo) [reply ui/mirage/config.js]: Good catch! This was a merge conflict that had been resolved incorrectly when main was last merged into this branch

**Review by hashishaw — APPROVED** [review]
